### PR TITLE
EVG-14861: refactor packages to flatten interface definitions

### PR DESCRIPTION
**Pull request by Kimchelly**
Jira: https://jira.mongodb.org/browse/EVG-14861

This is long, but I pretty much just refactored the interfaces and implementations into different locations. This will make testing a bit easier. I'll make a separate PR to actually implement the mock Secrets Manager client.

* Put interfaces in top-level `cocoa` package so they can be shared amongst the other packages without causing dependency cycles.
* Put implementations in subpackages (`secret`, `ecs`).
* Move some packages into `internal` because we don't really want to expose them in the public API.
* Make `internal/testutil` package for internal testing helpers.
* Rename a couple functions.

### CODE REVIEW

File: mock/vault.go [hunk 1]
```diff
@@ -4,7 +4,7 @@ import (
 	"context"
 	"errors"
 
-	"github.com/evergreen-ci/cocoa/secret"
+	"github.com/evergreen-ci/cocoa"
 )
 
 // Vault provides a mock implementation of a secret.Vault. This makes it
```
Comment (julianedwards) [inline mock/vault.go:10]: `cocoa.Vault`

File: secret/secrets_manager_vault.go [hunk 2]
```diff
@@ -12,18 +13,18 @@ import (
 // BasicSecretsManager provides a Vault implementation backed by Amazon Secrets
```
Comment (julianedwards) [inline secret/secrets_manager_vault.go:13]: `cocoa.Vault`

Comment (julianedwards) [reply secret/secrets_manager_vault.go:13]: Also, will it always be an amazon secrets manager?

File: doc.go [hunk 3]
```diff
@@ -3,14 +3,22 @@ Package cocoa provides interfaces to interact with groups of containers (called
...
-directly to the ECS API instead.
+The ECSClient provides a convenience wrapper around the ECS API. If the
+ECSPodCreator and ECSPod do not fulfill your needs, you can instead make calls
+directly to the ECS API using this client.
+
+The Vault is an ancillary service for pods that supports interacting with a
+dedicated secrets management service. It conveniently integrates with pods to
+securely pass secrets into containers.
+
+The SecretsManagerClient provides a convenience wrapper around the Secrets
```
Comment (julianedwards) [inline doc.go:20]: Is this AWS's secrets manager api? Maybe we should specify if it is.

File: ecs/pod_creator.go [hunk 4]
```diff
@@ -0,0 +1,37 @@
+package ecs
+
+import (
+	"context"
+
+	"github.com/evergreen-ci/cocoa"
+	"github.com/pkg/errors"
+)
+
+// BasicECSPodCreator provides an ECSPodCreator implementation to create
```
Comment (julianedwards) [inline ecs/pod_creator.go:10]: `cocoa.ECSPodCreator`

File: ecs/client.go [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/ecs"
+	"github.com/evergreen-ci/cocoa/internal/awsutil"
+	"github.com/evergreen-ci/utility"
+	"github.com/mongodb/grip"
+	"github.com/mongodb/grip/message"
+	"github.com/pkg/errors"
+)
+
+// BasicECSClient provides an ECSClient implementation that wraps the ECS API.
```
Comment (julianedwards) [inline ecs/client.go:18]: `cocoa.ECSClient`

File: doc.go [hunk 6]
```diff
@@ -3,14 +3,22 @@ Package cocoa provides interfaces to interact with groups of containers (called
 pods) backed by container orchestration services. Containers are not managed
 individually - they're managed as logical groupings of containers.
 
-The ECSPodCreator interface provides an abstraction to create pods backed by ECS
-without needing to make direct calls to the API.
+The ECSPodCreator provides an abstraction to create pods backed by ECS without
```
Comment (julianedwards) [inline doc.go:6]: Can we specify AWS ? `AWS ECS`.

File: secrets_manager_client.go [hunk 7]
```diff
@@ -0,0 +1,23 @@
+package cocoa
+
+import (
+	"context"
+
+	"github.com/aws/aws-sdk-go/service/secretsmanager"
+)
+
+// SecretsManagerClient provides a common interface to interact with a client
+// backed by Secrets Manager. Implementations must handle retrying and backoff.
```
Comment (julianedwards) [inline secrets_manager_client.go:10]: `AWS Secrets Manager`

**Review by julianedwards — CHANGES_REQUESTED** [review]
Few nits for clarifying. But I think it would be useful to specify, in the documentation, AWS before any services we are wrapping or interacting with. For example, `AWS Secrets Manager` or `AWS ECS`.

File: doc.go (hunk 3 above)
Comment (Kimchelly) [reply doc.go]: Sure, I can mention AWS (I also did it for ECS).

File: ecs/client.go (hunk 5 above)
Comment (Kimchelly) [reply ecs/client.go]: Fixed.

File: ecs/pod_creator.go (hunk 4 above)
Comment (Kimchelly) [reply ecs/pod_creator.go]: Done.

File: mock/vault.go (hunk 1 above)
Comment (Kimchelly) [reply mock/vault.go]: Fixed.

File: secret/doc.go [hunk 8]
```diff
@@ -1,13 +1,5 @@
 /*
 Package secret provides interfaces to interact with ancillary secrets management
```
[the remark this answers is not in the document]
Comment (Kimchelly) [reply secret/doc.go]: Sure, that's probably more accurate.

File: secret/secrets_manager_vault.go (hunk 2 above)
Comment (Kimchelly) [reply secret/secrets_manager_vault.go]: Yes, this implementation is backed by a Secrets Manager client.

File: secrets_manager_client.go (hunk 7 above)
Comment (Kimchelly) [reply secrets_manager_client.go]: Fixed.

File: doc.go (hunk 6 above)
Comment (Kimchelly) [reply doc.go]: Done.

**Review by julianedwards — APPROVED** [review]

**Pull request merged**